src/Microsoft.DotNet.Interactive.Tests/LanguageKernelPackageTests.cs

```diff
@@ -116,41 +116,93 @@ public async Task it_can_load_assembly_references_using_r_directive_separate_sub
         [InlineData(Language.FSharp, false)]
         [InlineData(Language.CSharp, true)]
         //[InlineData(Language.FSharp, true, Skip = "FSharp, relative is not relative to cwd.")]
-        public async Task it_can_load_assembly_references_using_r_directive_with_relative_path(Language language, bool changeWorkingDirectory)
+        public async Task it_can_load_assembly_references_using_r_directive_with_relative_path(Language language, bool changeCurrentDirectoryInUserCode)
         {
-            var workingDirectory = Directory.GetCurrentDirectory();
-            DisposeAfterTest(() => Directory.SetCurrentDirectory(workingDirectory));
+            var currentDirectory = Directory.GetCurrentDirectory();
+            DisposeAfterTest(() => Directory.SetCurrentDirectory(currentDirectory));
 
             var kernel = CreateKernel(language);
 
-            if (changeWorkingDirectory)
+            if (changeCurrentDirectoryInUserCode)
             {
+                //Even when a user changes the current directory, loading from a relative path is not affected.
                 await kernel.SendAsync(new SubmitCode("System.IO.Directory.SetCurrentDirectory(\"..\")"));
             }
 
-            var fullName = new FileInfo(typeof(JsonConvert).Assembly.Location).FullName;
+            var dllName = new FileInfo(typeof(JsonConvert).Assembly.Location).Name;             
 
-            var currentDirectoryName = new DirectoryInfo(Directory.GetCurrentDirectory()).Name;
+            var code = language switch
+            {
+                Language.CSharp => $"#r \"{dllName}\"",
+                Language.FSharp => $"#r \"{dllName}\""
+            };
 
-            var relativeDllPath = Path.GetRelativePath(
-                Directory.GetCurrentDirectory(),
-                fullName);
+            var command = new SubmitCode(code);
 
-            var relativePath =
-                Path.Combine(
-                    "..",
-                    currentDirectoryName,
-                    relativeDllPath)
-                .Replace("\\", "/");
+            await kernel.SendAsync(command);
 
-            var code = language switch
+            KernelEvents.Should()
+                        .ContainSingle<CommandSucceeded>(c => c.Command == command);
+        }
+
+        [Theory]
+        [InlineData(Language.CSharp, false)]
+        [InlineData(Language.FSharp, false)]
+        [InlineData(Language.CSharp, true)]
+        public async Task it_can_load_script_files_using_load_directive_with_relative_path(Language language, bool changeCurrentDirectoryInUserCode)
```

Comment:
Can we add a `changeCurrentDirectoryInUserCode == true` case for F#?

Reply:
I'll also put an IKernelCommandHandler for ChangeWorkingDirectory into FSharpKernel.fs because this also needs to be handled on that side.

I couldn't figure out how to change the working directory in FSharp-Script. That's why FSharp is commented for this test: "it_can_load_script_files_using_load_directive_with_relative_path_after_command_changeWorkingDirectory"

Is there somebody else who can have a look at the FSharp-Part?
Unfortunately I don't know a lot about FSharp yet.

Reply:
I was hesitating when it came to renaming "WorkingDirectory" because I was worrying to break the Contract to the Typescript-world, I don't know how you are auto-generating the Typescript-Contract and what I would break on this side.
But finally I renamed "WorkingDirectory" to "WorkspaceDirectory" where appropriate.
The CSharpKernel uses an "InteractiveWorkspace" so this seemed like a good choice to me.

The VSCode-Add-In still works after that change. I believe the Command "ChangeWorkingDirectory" was not sent from there so far.

Reply:
Sorry, I was being unclear. We shouldn't change the JSON API contracts. As long as we don't rename the command, we avoid that. I was really trying to work out what a better name for this concept would be and we can plan a contract change for later if we come up with something we like. 

For F# help, we can summon @KevinRansom or @brettfo. For the TypeScript-y bits, we can summon @brettfo.

Reply:
I'm not sure what the F# part is to update it's probing paths for `#load` so I'll throw this one over to @KevinRansom.

Reply:
It's not just **#load** it's also **#r**, and I suppose **#i**.  @dsyme has been pushing a PR that addresses a **#r** relative load issue that's related to TypeProviders, and I have been maliciously sitting on it.  Because I wanted to think about it some.

Historically **#r** and **#load** in F# has been relative to the actual script in which it is situated.  This is actually the cleanest approach, it makes scripts with relative references more reusable.  The executing script is not always a file on disk, but an in memory file, often part of a notebook file, this is pretty much identical to standard FSI though, where the executing script is not on disk but treated as if it were, and in the current working directory when FSI starts.  There are circumstances when the executing script is a file generated by the scripting infrastructure, E.g. the list of references generated in a **#r "nuget: ..."**, that script is generated in the temp directory with a random number in the path name and, if I remember correctly a thread.id so relative to that is really not a thing.

I would really ... really like to consider all of the scenarios, before blindly updating fsi to change how #load happens.  Changing FSI now would break existing scripts, that have worked this way on released FSI for 13 years, you can probably guess that I would be reluctant to make that change.  However, the way, one should probably think about this issue is that:

**#r**, **#i** and **#load** are relative to the location for the file in which they are found, and the "project directory(working directory) or however you want to think about is the path for in-memory scripts executed within the notebook.

So for a notebook: **//user/kevinr/notebooks/kevinsfirst/kevinsfirst.dib**

#load "compilationscripts/somefile.fsx" will load from:
**//user/kevinr/notebooks/kevinsfirst/compilationscripts/somefile.fsx**

And if that file has a line:
#load "codegen/makeframework.fsx"  will load from:
**//user/kevinr/notebooks/kevinsfirst/compilationscripts/codegen/makeframework**

This does leave the issue of what to do about the nuget / typeprovider scenario, @dsyme has a proposal and a PR, I am not certain it is exactly right but his proposals are mostly spot on, so I expect it fine will be when I finally try to convince myself.